### ISSUE TITLE
CPU maxes out with jack2 + worker threads above 1 and unlocked

### ISSUE THREAD

**Issue by milkmiruku** [j3ffhubb/musikernel#8]
On Arch Linux 4.4.6-ck (BFS), using `musikernel-git` in AUR, first time running (and then removing the config directory between tries) and everything (including the mouse pointer) freezes right after the GUI appears, requiring a hard reboot. I tried to redirect stdout and stderr to a logfile, but, from doing the same with ALSA, I think these only get written to when the program quits normally.

**Comment by milkmiruku** [comment]
Tried 4.1.20-1-lts, it launches but after a few seconds it maxes out my two core system. stop_server came after Ctrl-q y;

> Detected Intel CPU, checking for hyperthreading.
> Spawning 2 worker threads
>  Attempting to set .__sched_priority = 99
> Scheduling successfully applied with priority 99
>  Actual output latency:
> 	milliseconds:  23.219954
> 	samples:  1024
> Engine sent 'ready' message
> stop_server called
> PYDAW_SUBPROCESS did not exit on it's own, sending SIGKILL...
> Successfully processed all pending events in 3 iterations
> Successfully garbage collected all objects
> UI process doesn't exist, exiting.

but the non-root backend process continued hogging CPU and required a manual kill.

**Comment by j3ffhubb** [comment]
Hello,

Thanks for reporting this.  I have a couple of questions, just to make sure I properly understand.

1. So, you're saying that JACK does work and outputs audio as long as the thread count is set to 1?  Unfortunately, that might be an expected behavior on a Core2Duo when using JACK, given that JACK runs in it's own process, and MusiKernel has 2 threads, they are probably blocking and deadlocking each other, and Core2 doesn't have any hacks like Hyperthreading that might allow it do both at the same time.  What I can do (and probably will do) is go a bit more conservative on how many threads to spawn in "Auto" mode, but the end result will be using 1 thread when a dual-core is detected.
2. With ALSA, are you sure that you have the correct output device selected?  I personally use ALSA and have never really had problems with it on any PC, but you won't hear any audio unless the correct device is selected.  Also, if your audio device has more than 2 outputs, you might have have to change which outputs it's using in the Transport->Menu->File->HardwareSettings->DefaultOutputs.  ALSA runs in kernel mode and not in a separate userspace daemon, so it's not surprising that 2 threads run OK here based on my theory.

**Comment by j3ffhubb** [comment]
I didn't mean to close the issue before you confirmed that everything was working, Github did that automatically when I added "Fixes #8" to the commit :-)

**Comment by milkmiruku** [comment]
Apologies, the text I edited in for the second working scenario was ambiguous; JACK mode works with "2" or "Auto" Worker Threads when the 'Lock worker threads to own core?" option is enabled.

My audio setup is usually Cadence with the "ALSA -> PulseAudio -> JACK" bridge.

tl;dr - the "ALSA -> PulseAudio -> JACK" bridge in Cadence doesn't work with MusiKernel, but the "ALSA -> JACK" bridge mode does.

The following is more a note than an issue :)

I usually use the APJ version so I can easily control everything in pavucontrol. When in this bridge mode, the MK device configuration dialog window, with ALSA Host API selected, shows Output Device options of "default", "M Audio Audiophile 23/96: ICE 1712 multi (hw:1,0)" and "pulse".

Trying with "default" or "pulse", after adding an audio file to a track, hitting play does nothing. Selecting "M Audio.." does play, but I don't have the card hooked up to anything atm.

To avoid confusion, here's the output from launching in a terminal;

> ...
> ALSA lib pcm_dsnoop.c:606:(snd_pcm_dsnoop_open) unable to open slave
> ALSA lib pcm_dmix.c:1029:(snd_pcm_dmix_open) unable to open slave
> ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
> ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
> ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
> ALSA lib pcm_dmix.c:1029:(snd_pcm_dmix_open) unable to open slave
> 
> Device Index: 0
> Name : HDA Intel: ALC883 Alt Analog (hw:0,2)
> maxInputChannels : 2
> maxOutputChannels : 0
> 
> Device Index: 1
> Name : M Audio Audiophile 24/96: ICE1712 multi (hw:1,0)
> maxInputChannels : 12
> maxOutputChannels : 10
> 
> Device Index: 2
> Name : USB Device 0x46d:0x8ad: Audio (hw:2,0)
> maxInputChannels : 1
> maxOutputChannels : 0
> 
> Device Index: 3
> Name : pulse
> maxInputChannels : 32
> maxOutputChannels : 32
> 
> Device Index: 4
> Name : default
> maxInputChannels : 128
> maxOutputChannels : 128
> 
> Device Index: 5
> Name : system
> maxInputChannels : 2
> maxOutputChannels : 2
> 
> Device Index: 6
> Name : PulseAudio JACK Sink
> maxInputChannels : 2
> maxOutputChannels : 0
> 
> Device Index: 7
> Name : PulseAudio JACK Source
> maxInputChannels : 0
> maxOutputChannels : 2

If I play something in alsaplayer, it is routed to PulseAudio, and, on the Playback tab of pavucontrol, it's output device is set to "Jack sink (PulseAudio JACK Sink)"

(And FYI, just to round things off, when selecting JACK as Host API, I get the options "PulseAudio JACK Source" and "system". The latter is the one that works (it's the master out). If I select the former, it plays but I get no output, but this is expected as it's outputting from JACK to PA -  "PulseAudio JACK Source" shows up on Input Devices in pavucontrol, and, though I'm not monitoring it, I'd expect to see a graphical level moving. This is just curious though and the least of my worries)

The wonders of Linux audio system abstraction :)

**Comment by j3ffhubb** [comment]
You hit the nail on the head: `The wonders of Linux audio system abstraction` :-)

I think I understand why Pulseaudio doesn't work.  MusiKernel calls 'pasuspender' if it's found in $PATH when it launches, which then makes Pulseaudio release all of it's resources until MusiKernel closes.  I could maybe make the call to pasuspender optional, but I'd need to investigate further.

Most of those configuration options only exist in Linux (for a reason), on the Mac and Windows ports you wouldn't even see them.  I obviously love Linux (former Red Hat employee), but it often is a matter of tweaking the configuration until it works.

Some thoughts and ideas on settings and performance tuning:
- MusiKernel's plugins are very CPU efficient, being an all-in-one system eliminates the overhead of foreign function calls to .so's, and allows end-to-end optimization by the compiler;  If one core can provide enough power to run your plugins, then that's actually preferable from a latency perspective over having to synchronize multiple threads.
- MusiKernel uses Portaudio to talk to each of the audio APIs, so there's not much that can be done to fix problems with particular audio APIs (assuming the bugs aren't in the API themselves), it would have to be fixed in Portaudio.
- JACK not working with the threads not locked to cores actually sounds like a kernel bug with your system, if anything I'd expect that bug to happen with threads locked (the potential for deadlocking is the reason why threads locked isn't the default)

**Comment by j3ffhubb** [comment]
I'm closing this issue, I believe that the more conservative thread count when set to "auto" is all that can reasonably be done to fix the issue.  Given that this is almost certainly a kernel scheduler issue (and possibly a JACK issue), there's not much else that could be done to detect or remediate these conditions at run-time, from user-space.
